scrapy/settings/__init__.py

```diff
@@ -45,64 +64,224 @@ def __str__(self):
+        :type default: any
         """
         return bool(int(self.get(name, default)))
 
     def getint(self, name, default=0):
+        """
+        Get a setting value as an int.
+
+        :param name: the setting name
+        :type name: string
+
+        :param default: the value to return if no setting is found
+        :type default: any
+        """
         return int(self.get(name, default))
 
     def getfloat(self, name, default=0.0):
+        """
+        Get a setting value as a float.
+
+        :param name: the setting name
+        :type name: string
+
+        :param default: the value to return if no setting is found
+        :type default: any
+        """
         return float(self.get(name, default))
 
     def getlist(self, name, default=None):
+        """
+        Get a setting value as a list. If the setting original type is a list, a
+        copy of it will be returned. If it's a string it will be split by ",".
+
+        For example, settings populated through environment variables set to
+        ``'one,two'`` will return a list ['one', 'two'] when using this method.
+
+        :param name: the setting name
+        :type name: string
+
+        :param default: the value to return if no setting is found
+        :type default: any
+        """
         value = self.get(name, default or [])
         if isinstance(value, six.string_types):
             value = value.split(',')
         return list(value)
 
     def getdict(self, name, default=None):
+        """
+        Get a setting value as a dictionary. If the setting original type is a
+        dictionary, a copy of it will be returned. If it is a string it will be
+        evaluated as a JSON dictionary. In the case that it is a
+        :class:`~scrapy.settings.BaseSettings` instance itself, it will be
+        converted to a dictionary, containing all its current settings values
+        as they would be returned by :meth:`~scrapy.settings.BaseSettings.get`,
+        and losing all information about priority and mutability.
+
+        :param name: the setting name
+        :type name: string
+
+        :param default: the value to return if no setting is found
+        :type default: any
+        """
         value = self.get(name, default or {})
         if isinstance(value, six.string_types):
             value = json.loads(value)
         return dict(value)
 
+    def _getcomposite(self, name):
+        # DO NOT USE THIS FUNCTION IN YOUR CUSTOM PROJECTS
+        # It's for internal use in the transition away from the _BASE settings and
+        # will be removed along with _BASE support in a future release
+        basename = name + "_BASE"
+        if basename in self:
+            warnings.warn('_BASE settings are deprecated.',
+                          category=ScrapyDeprecationWarning)
+            compsett = BaseSettings(self[name + "_BASE"], priority='default')
+            compsett.update(self[name])
+            return compsett
+        else:
```

Comment:
Now that I've seen it with a little more distance, I think this whole function doesn't really achieve what I had intended.

When users override `XY_BASE`, they explicitly don't want any of Scrapy's defaults for that component setting. But line 206 pulls Scrapy's defaults (which now live in `XY`) back in, and even worse overwrites the users `XY_BASE` settings where they have the same keys (say when the user simply changed some orders).

I guess either line 206 could be changed so that only those keys from `XY`that have a priority higher than `default` are considered, or support for `_BASE` settings could be dropped altogether with this PR. After all, they have always been marked as "never edit this", there is going to be a release note b/c of the new update-not-overwrite behaviour of the dict-like settings anyways, and we could get rid of this non-public helper function.